projects/power_distribution/src/publish_data.c

```diff
@@ -19,17 +30,42 @@ StatusCode power_distribution_publish_data_init(PowerDistributionPublishConfig c
   return STATUS_CODE_OK;
 }
 
+static void prv_partially_publish(SoftTimerId timer_id, void *context) {
+  if (s_current_measurements == NULL) {
+    // guard against possible segfaults
+    return;
+  }
+
+  // We use a uintptr_t for the index so we can pass it through the context
+  uintptr_t index = (uintptr_t)context;
+
+  uint16_t transmit_up_to = MIN(index + NUM_MSG_PER_ITERATION, s_config.num_currents_to_publish);
+  while (index < transmit_up_to) {
+    PowerDistributionCurrent current_id = s_config.currents_to_publish[index];
+    StatusCode tx_code = s_config.transmitter(current_id, s_current_measurements[current_id]);
+    if (!status_ok(tx_code)) {
+      LOG_WARN("Failed to TX current %d: status code %d\n", current_id, tx_code);
+    }
+    index++;
+  }
+
+  if (index < s_config.num_currents_to_publish) {
+    StatusCode code =
+        soft_timer_start_millis(ITERATION_WAIT_MS, prv_partially_publish, (void *)index, NULL);
+    if (!status_ok(code)) {
+      LOG_WARN("Failed to start publish_data soft timer! status code: %d\n", code);
+    }
+  }
+}
+
 StatusCode power_distribution_publish_data_publish(
     uint16_t current_measurements[NUM_POWER_DISTRIBUTION_CURRENTS]) {
   if (!s_config.currents_to_publish) {
     return status_code(STATUS_CODE_UNINITIALIZED);
   }
 
-  // transmit each current from the config
-  for (uint16_t c = 0; c < s_config.num_currents_to_publish; c++) {
-    PowerDistributionCurrent current_id = s_config.currents_to_publish[c];
-    status_ok_or_return(s_config.transmitter(current_id, current_measurements[current_id]));
-  }
-
+  s_current_measurements = current_measurements;
+  uintptr_t index = 0;
```

Comment:
What are the benefits of using a uintptr as opposed to a uint8_t by reference? Is it just so you don't have to cast it to use it?

Reply:
If we used a uint8_t by reference we'd have to either malloc it (possible memory leaks) or make it a static variable, since it can't live on the stack since any function might return before the soft timer is called. If it's a static variable, then if we called `power_distribution_publish_data_publish` twice before the all the currents were tx'd, the counter might get messed up and messages might not get sent. With a uintptr_t we can just stuff the counter into the pointer itself and avoid these issues.